defaultNatsService.go

```diff
@@ -0,0 +1,346 @@
+package natsio
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/nats-io/nats.go"
+	"github.com/orchestd/dependencybundler/interfaces/configuration"
+	"github.com/orchestd/dependencybundler/interfaces/credentials"
+	"github.com/orchestd/dependencybundler/interfaces/log"
+	. "github.com/orchestd/servicereply"
+	"github.com/orchestd/servicereply/status"
+	"go.uber.org/fx"
+	"reflect"
+	"time"
+)
+
+var connect = func(natsUrl string, options ...nats.Option) (NatsConnection, error) {
+	return nats.Connect(natsUrl, options...)
+}
+
+func NewDefaultNatsService(lc fx.Lifecycle, config configuration.Config, logger log.Logger, credentials credentials.CredentialsGetter) NatsService {
+	return getDefaultService(lc, config, logger, func(serviceName string) nats.Option {
+		natsUser := credentials.GetCredentials().NatsUser
+		if natsUser == "" {
+			panic("NotificationService could not get credentials by key NatsUser")
+		}
+		natsPw := credentials.GetCredentials().NatsPw
+		if natsPw == "" {
+			panic("NotificationService could not get credentials by key NatsPw")
+		}
+		authOpt := nats.UserInfo(natsUser, natsPw)
+		return authOpt
+	})
+}
+
+func NewDefaultNatsServiceWithJWTAuth(lc fx.Lifecycle, config configuration.Config, logger log.Logger) NatsService {
+	return getDefaultService(lc, config, logger, func(serviceName string) nats.Option {
+		authOpt := nats.UserCredentials(serviceName + ".creds")
+		return authOpt
+	})
+}
+
+func getDefaultService(lc fx.Lifecycle, config configuration.Config, logger log.Logger, getAuthOpts func(serviceName string) nats.Option) NatsService {
+	service := defaultNatsService{
+		logger:        logger,
+		subscriptions: map[string]*nats.Subscription{},
+	}
+	serviceName, err := config.GetServiceName()
+	if err != nil {
+		panic("can't get serviceName: " + err.Error())
+	}
+	natsUrl, err := config.Get("NatsUrl").String()
+	if err != nil {
+		panic("could not get conf by key NatsUrl " + err.Error())
+	}
+	connectionAttempts, err := config.Get("connectionAttempts").Int()
+	if err != nil {
+		connectionAttempts = 1
+	}
+	reconnectionAttempts, err := config.Get("reconnectionAttempts").Int()
+	if err != nil {
+		reconnectionAttempts = -1
+	}
+	authOpt := getAuthOpts(serviceName)
+	lc.Append(fx.Hook{
+		OnStart: func(c context.Context) error {
+			return service.Connect(natsUrl, serviceName, authOpt, connectionAttempts, reconnectionAttempts)
+		},
+		OnStop: func(c context.Context) error {
+			service.Close()
+			return nil
+		},
+	})
+	return &service
+}
+
+type defaultNatsService struct {
```

Comment:
default... unlike - none default ? do we support manually created none default ?

Reply:
default - custom implementation of NatsService interface.

Reply:
do we support manually created none default ?